shared/modules/Numeric.ts

```diff
@@ -0,0 +1,621 @@
+ * The only valid strings for this method are those that are either hexadecimal
+ * numeric values OR numeric strings that can be converted to BigNumbers. It is
+ * impossible to tell the difference between a hex value of 100000 vs a decimal
+ * value of 100000 so a second parameter indicating the numeric base of the
+ * string value must be provided.
+ *
+ * @param value - A hexadecimal or decimal string
+ * @param numericBase - Either 16 for a hexadeciaml or 10 for a decimal
+ * @returns A BigNumber representation of the value
+ */
+function stringToBigNumber(value: string, numericBase: NumericBase) {
+  if (typeof value !== 'string') {
+    throw new Error(
+      `Value of type ${typeof value} passed to stringToBigNumber`,
+    );
+  }
+  if (
+    numericBase === 16 &&
+    (isHexStringOrNegatedHexString(value) || isDecimalHex(value))
+  ) {
+    return hexadecimalToBigNumber(value);
+  } else if (
+    numericBase === 10 &&
+    // check if we have a finite integer or float
+    (isFinite(parseInt(value, 10)) || isFinite(parseFloat(value)))
+  ) {
+    return decimalToBigNumber(value);
+  }
+  throw new Error(
+    `String provided to stringToBigNumber is not a hexadecimal or decimal string: ${value}, ${numericBase}`,
+  );
+}
+
+/**
+ * This method will convert a hexadecimal or deciaml number into a BigNumber.
+ * The second parameter must be supplied and determines whether to treat the
+ * value as a hexadecimal or decimal value.
+ *
+ * @param value - hexadecimal or decimal number[]
+ * @param numericBase - 10 for decimal, 16 for hexadecimal
+ * @returns BigNumber representation of the value
+ */
+function numberToBigNumber(value: number, numericBase: NumericBase) {
+  if (typeof value !== 'number') {
+    throw new Error(
+      `Value of type ${typeof value} passed to numberToBigNumber`,
+    );
+  }
+  if (numericBase === 16 && isHexString(`${value}`)) {
+    return new BigNumber(`${value}`, 16);
+  }
+  return new BigNumber(value, 10);
+}
+
+/**
+ * Method to convert a BN to a BigNumber
+ *
+ * @param value - A BN representation of a value
+ * @returns A BigNumber representation of the BN's underlying value
+ */
+function bnToBigNumber(value: BN) {
+  if (value instanceof BN === false) {
+    throw new Error(
+      `value passed to bnToBigNumber is not a BN. Received type ${typeof value}`,
+    );
+  }
+  return new BigNumber(value.toString(16), 16);
+}
+
+/**
+ * Converts a value of the supported types (string, number, BN) to a BigNumber.
+ *
+ * @param value - The value to convert to a BigNumber
+ * @param numericBase - The numeric base of the underlying value
+ * @returns A BigNumber representation of the value
+ */
+function valueToBigNumber(value: string | number, numericBase: NumericBase) {
+  if (typeof value === 'string') {
+    return stringToBigNumber(value, numericBase);
+  } else if (typeof value === 'number' && isNaN(value) === false) {
+    return numberToBigNumber(value, numericBase);
```

Comment:
A string could be a base 10 string like '10' or a base 16 string like 'a', it could also be a something like '0x-a' or '0.a7' the latter of these would error if we attempt to cast to a number. Furthermore for very large numbers represented by hexadecimal casting to number would cause issues with accuracy of the number. Javascript can only accurately and safely manage numbers to a certain value.

A Number might be NaN and casting it to string would result in errors. Your suggestion makes sense but for the sake of treating each possible special case i think it would be best if we deal with them separately.

Reply:
Thanks for the clear explanation, now i can see different cases we need to deal with in each side, makes sense to separate definitely 👍